src/providers/bibtexformater.ts

```diff
@@ -67,56 +96,75 @@ export class BibtexFormater {
+                        entryLocations[i].start.line + lineDelta + lineOffset,
+                        entryLocations[i].start.character + columnOffset,
+                        entryLocations[i].start.line + lineDelta + (sortedEntryLocations[i].end.line - sortedEntryLocations[i].start.line) + lineOffset,
                         entryLocations[i].end.character
                     )
-                    diags.push(new vscode.Diagnostic(
+                    this.diags.push(new vscode.Diagnostic(
                         highlightRange,
                         `Duplicate entry "${entries[i].internalKey}".`,
                         vscode.DiagnosticSeverity.Warning
                     ))
                 } else { // 'Comment Duplicates'
                     // Log duplicate entry since we aren't highlighting it
                     this.extension.logger.addLogMessage(
-                        `BibTeX-format: Duplicate entry "${entries[i].internalKey}" at line ${entryLocations[i].start.line + lineDelta + 1}.`)
+                        `BibTeX-format: Duplicate entry "${entries[i].internalKey}" at line ${entryLocations[i].start.line + lineDelta + 1 + lineOffset}.`)
                     text = text.replace(/@/,'')
                 }
             }
 
             // Put text from entry[i] into (sorted)location[i]
-            edit.replace(uri, entryLocations[i], text)
+            edits.push(new vscode.TextEdit(new vscode.Range(entryLocations[i].start.translate(range?.start.line, range?.start.character), entryLocations[i].end.translate(range?.start.line)), text))
 
             // We need to figure out the line changes in order to highlight properly
             lineDelta += (sortedEntryLocations[i].end.line - sortedEntryLocations[i].start.line) - (entryLocations[i].end.line - entryLocations[i].start.line)
         }
+        this.extension.logger.addLogMessage('Formatted ' + document.fileName)
+        return edits
+    }
+}
 
-        vscode.workspace.applyEdit(edit).then(success => {
-            if (success) {
-                this.duplicatesDiagnostics.set(uri, diags)
-                const t1 = performance.now()
-                this.extension.logger.addLogMessage(`BibTeX action successful. Took ${t1 - t0} ms.`)
-            } else {
-                this.extension.logger.showErrorMessage('Something went wrong while processing the bibliography.')
-            }
-        })
+export class BibtexFormatterProvider implements vscode.DocumentFormattingEditProvider, vscode.DocumentRangeFormattingEditProvider {
+    private formatter: BibtexFormatter
+    extension: Extension
+
+    constructor(extension: Extension) {
+        this.extension = extension
+        this.formatter = new BibtexFormatter(extension)
+    }
+
+    public provideDocumentFormattingEdits(document: vscode.TextDocument, _options: vscode.FormattingOptions, _token: vscode.CancellationToken): vscode.ProviderResult<vscode.TextEdit[]> {
```

Comment:
We should make `provideDocumentFormattingEdits` async.

Reply:
It does not seem possible to declare a function with return type `vscode.ProviderResult<T>` async. We would need to change the return type to `Promise<T>` but https://code.visualstudio.com/api/references/vscode-api#DocumentFormattingEditProvider

Btw, we do not use `async` for the latex formatter 
https://github.com/James-Yu/LaTeX-Workshop/blob/2ad71efb30fdd6a9b03f4ab3d7bb24b567928cc2/src/providers/latexformatter.ts#L210-L219

Reply:
What I am saying is that the following is easier to read and maintain:
```ts
    public async provideDocumentFormattingEdits(document: vscode.TextDocument, _options: vscode.FormattingOptions, _token: vscode.CancellationToken) {
        const edits = await this.formatter.formatDocument(document, true, true)
        this.formatter.duplicatesDiagnostics.set(document.uri, this.formatter.diags)
        this.extension.logger.addLogMessage('BibTeX action successful.')
        return edits
    }
```

Reply:
I feel a bit lost here. According to  https://code.visualstudio.com/api/references/vscode-api#DocumentFormattingEditProvider, `provideDocumentFormattingEdits` must have return type `vscode.ProviderResult<T>`, which is not possible for an `async` function.

Reply:
With the diagnostics removed (see 6d27c1b), I think it is fine now?